### ISSUE TITLE
Failed to work with async_trait

### ISSUE THREAD

**Issue by linxiulei** [xacrimon/dashmap#55]
Not sure this is right place to file the issue, but async_trait works fine with std::collections::HashMap, so allow me to post here first. I tried to build following code

```
use std;
use async_trait::async_trait;
use dashmap::DashMap;
use std::collections::HashMap;
use tokio;

struct Foo {}
struct Gar<T: Bar> {
    map: DashMap<u32, T>,
}

#[async_trait]
trait Bar {
    async fn bar(&self) -> u32;
}

#[async_trait]
impl Bar for Foo {
    async fn bar(&self) -> u32{
        0
    }
}

#[async_trait]
trait GarTrait {
    async fn coo(&self) -> Vec<u32>;
}

#[async_trait]
impl<T> GarTrait for Gar<T> where T: Bar + Send + Sync,{
    async fn coo(&self) -> Vec<u32> {
        let mut a = Vec::new();
        for item in self.map.iter() {
            let bar= item.value();
            let i: u32 = bar.bar().await;
            a.push(i);
        }
        a
    }
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>>{
    let g = Gar::<Foo>{map:DashMap::new()};
    g.coo().await;
    Ok(())
}
```

`cargo build` complaints error 

```
$ cargo build
   Compiling t v0.1.0 (-)
warning: unused import: `std::collections::HashMap`
 --> src/main.rs:4:5
  |
4 | use std::collections::HashMap;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default

error[E0311]: the parameter type `T` may not live long enough
  --> src/main.rs:31:37
   |
31 |       async fn coo(&self) -> Vec<u32> {
   |  _____________________________________^
32 | |         let mut a = Vec::new();
33 | |         for item in self.map.iter() {
34 | |             let bar= item.value();
...  |
38 | |         a
39 | |     }
   | |_____^
   |
   = help: consider adding an explicit lifetime bound for `T`
   = note: the parameter type `T` must be valid for any other region...
note: ...so that the type `T` will meet its required lifetime bounds
  --> src/main.rs:31:37
   |
31 |       async fn coo(&self) -> Vec<u32> {
   |  _____________________________________^
32 | |         let mut a = Vec::new();
33 | |         for item in self.map.iter() {
34 | |             let bar= item.value();
...  |
38 | |         a
39 | |     }
   | |_____^

error: aborting due to previous error

error: could not compile `t`.

To learn more, run the command again with --verbose.
```

Note: It's ok to compile when using self.map.get()

Cargo.toml

```
[dependencies]
tokio = { version = "*", features = ["full"] }
async-trait = "*"
dashmap = "*"
```

**Comment by xacrimon** [comment]
First of all, I would suggest fixing your dependency versions. Wildcard versions are dangerous and may bite you later.

This is due to how boxed trait objects interact with lifetimes. Async trait does this by appending the `'async_trait` lifetime where needed. In your function, this should be fixed by constraining T with that lifetime by adding `+ 'async_trait` to it's where bound.

**Comment by xacrimon** [comment]
This is really an issue with how async_trait works around lifetimes and not with dashmap.

**Comment by linxiulei** [comment]
Thanks for reply. You're probably right, though I failed with doing so.

**Comment by dtolnay** [comment]
@xacrimon please take a look at https://github.com/dtolnay/async-trait/issues/69#issuecomment-585466063. I don't think this is related to boxed trait objects, it's the Send impls in dashmap being stricter than they should be.

**Comment by xacrimon** [comment]
I just released v3.4.4 which fixes this!
